book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+comprises the whole ABI for any system components. For example, type includes
+structs, serializer logics and implemented protocol communication procedures.
+
+ABI item digest: Some hash based on its fields' type information. Used casual
+detection of unintended possible ABI change.
+
+ABI item revision: Monotonic increasing integer starting with 0, precisely
+identifying specific implementation at any range of time in the past or in the
+(near) future. All existing ABI item is unversioned and regarded as being
+revision 0.
+
+Versioning: Start adding variant of behaviors for given ABI item and starting to
+increment revision by 1.
+
+Revision list: List of revisions for given ABI item. Each list element has an
+associated range of time in some kind of units (epoch, cluster version, data
+format magic number) and other meta data for each revision like compatibility
+statuses. The time range could be overlapping or non-overlapping.
+
+Compatibility: Being able to process information from external source according
+to the specified behavior with the deterministic outcome.
+
+Compatibility status: What kind of strictness of system component should be
+desired when new (=not explicitly recognized) form of information is
+encountered. This can be classified as ignore/reject.
+
+Maintenance tiers: How incompatible changes should be introduced to the
+cluster. This means either (planned) hard-fork, urgent security fix requiring
+ABI change (Tier-1), or normal periodical voluntary software update (Tier-2).
+Tier-1 includes transactions and accounts, for example. Tier-2 includes shreds
+and snapshots, for example.
+
+Hard fork: A hard fork is a change to the cluster that makes previously invalid
+communications valid, and therefore requires all users / software to upgrade
+(definition is based on [1]). The upgrade is programmed and scheduled and takes
+place effective at specific future epoch.
+
+# Examples
+
+## Example 1: If we were adding #7233 after some time since mainnet launch:
+
+```patch
+@@applicaton code@@
+
+ #[solana_abi(allow=add, digest="79caeee9")]
+   // ^^ adding a member to allow=add enum doesn't trigger digest change
+ #[derive(SolanaABI)]
+ pub enum VoteError {
+     ....
++
++    #[error("vote timestamp not recent")]
++    TimestampTooOld,
+ }
+
+-#[solana_abi(allow=none, digest="3ffc9921")]
++#[solana_abi(allow=none, digest="1c6a53e9")]
+ #[derive(Serialize, Default, Deserialize, Debug, PartialEq, Eq, Clone)] 
+ pub struct Vote {
+     /// A stack of votes starting with the oldest vote
+     pub slots: Vec<Slot>,
+     /// signature of the bank's state at the last slot
+     pub hash: Hash,
++    #[solana_abi_added(since = "rev1")]
++    /// processing timestamp of last slot
++    pub timestamp: Option<UnixTimestamp>,
+ }
+
+ impl SolanaABI for Vote {
+   const REVISION_LIST = [
+      (0, "epoch", 0..), // leave this revision acceptable for indefinite time
++     (1, "epoch", 21..),
```

Comment:
Is this level of ABI management is too much grained? I think this can be justified by the finer and dense unit testing. But it could be regarded as an overkill. This is my preference, and there are seems no projects doing similar practice.

Reply:
Yes this does feel like overkill to me.

Reply:
Yeah, I've greatly stripped this down. :)